benches/nexmark/generator/person_generator.rs

```diff
@@ -0,0 +1,246 @@
+    Person {
+        id: last_base0_person_id(next_event_id) + FIRST_PERSON_ID,
+        name: next_person_name(rng),
+        email_address: next_email(rng),
+        credit_card: next_credit_card(rng),
+        city: next_us_city(rng),
+        state: next_us_state(rng),
+        date_time: DateTime::from_timestamp(secs.try_into().unwrap(), nsecs.try_into().unwrap()),
+        extra: String::new(),
+    }
+}
+
+/// Return a random person id (base 0).
+// TODO(absoludity): Update to take GeneratorConfig.
+pub fn next_base0_person_id<R: Rng + ?Sized>(event_id: Id, rng: &mut R) -> Id {
+    // Choose a random person from any of the 'active' people, plus a few 'leads'.
+    // By limiting to 'active' we ensure the density of bids or auctions per person
+    // does not decrease over time for long running jobs.
+    // TODO(absoludity): Understand why this code appears to shift the active
+    // people ids to always be the most recent people, rather than what the
+    // comment above claims.
+
+    // By choosing a person id ahead of the last valid person id we will make
+    // newPerson and newAuction events appear to have been swapped in time.
+    let num_people = last_base0_person_id(event_id) + 1;
+    let active_people = std::cmp::min(num_people, NUM_ACTIVE_PEOPLE);
+    let n = rng.gen_range(0..active_people);
+    num_people - active_people + n
+}
+
+/// Return the last valid person id (ignoring FIRST_PERSON_ID). Will be the
+/// current person id if due to generate a person.
+// TODO(absoludity): Update to take GeneratorConfig.
+pub fn last_base0_person_id(event_id: Id) -> Id {
+    let epoch = event_id / TOTAL_PROPORTION;
+    let mut offset = event_id % TOTAL_PROPORTION;
+
+    if offset >= PERSON_PROPORTION {
+        // About to generate an auction or bid.
+        // Go back to the last person generated in this epoch.
+        offset = PERSON_PROPORTION - 1;
+    }
+    // About to generate a person.
+    epoch * PERSON_PROPORTION + offset
+}
+
+// Return a random US state.
+fn next_us_state<R: Rng + ?Sized>(rng: &mut R) -> String {
+    US_STATES.choose(rng).unwrap().to_string()
+}
+
+// Return a random US city.
+fn next_us_city<R: Rng + ?Sized>(rng: &mut R) -> String {
+    US_CITIES.choose(rng).unwrap().to_string()
+}
+
+// Return a random person name.
+fn next_person_name<R: Rng + ?Sized>(rng: &mut R) -> String {
+    format!(
+        "{} {}",
+        FIRST_NAMES.choose(rng).unwrap(),
+        LAST_NAMES.choose(rng).unwrap()
+    )
+}
+
+// Return a random email address.
+fn next_email<R: Rng + ?Sized>(rng: &mut R) -> String {
+    format!("{}@{}.com", next_string(rng, 7), next_string(rng, 5))
+}
+
+// Return a random credit card number.
+fn next_credit_card<R: Rng + ?Sized>(rng: &mut R) -> String {
+    format!(
+        "{:04} {:04} {:04} {:04}",
+        rng.gen_range(0..10_000),
+        rng.gen_range(0..10_000),
+        rng.gen_range(0..10_000),
+        rng.gen_range(0..10_000)
+    )
+}
```

Comment:
This all would be much better suited as a struct that contains the rng for which these are methods on instead of free functions.
This can probably be done via a `Generator`/`DataGenerator`/`NextmarkData`/etc. struct within `generator.rs` and `impl` blocks on it within `person.rs` and `strings.rs`

Reply:
Hmm, I was keen to stay as close as possible to the structure of the original nexmark implementation which passes the Rng to the functions, even though they're java classes, as you can see in the equiv function here:

https://github.com/nexmark/nexmark/blob/54974ef36a0d01ef8ebc0b4ba39cfc50136af0f6/nexmark-flink/src/main/java/com/github/nexmark/flink/generator/model/PersonGenerator.java#L120-L130

It seems like the intention is to avoid maintaining state and keep it purely functional where possible (for eg, you'll also see the `next_person` function also takes the current event id as an arg, which could also be state of some `NextmarkData` struct).

Would you mind if I do this initial PR for the person following the functional approach of the java implementation, then I can do a quick follow-up PR switching to a struct-basd generator so it's easier to compare and discuss there?

Reply:
Sure, that's fine by me

Reply:
Done in https://github.com/absoludity/database-stream-processor/pull/1